### ISSUE TITLE
Generate JSON directly with SQL

### ISSUE THREAD

**Issue by mdempsky** [dxe/adb#62]
I think a lot of our SQL queries could be simplified by: (1) having the SQL query grab all the data we're interested in in one pass, and (2) letting MySQL handle aggregating the data into JSON.

Currently we spend a lot of Go code implementing this ourselves. E.g., here's a query that implements the entirety of GetEventsJSON, at least for one particular search predicate (`date >= '2019-12-01`).

```
select json_arrayagg(json_object(
    'id', id,
    'name', name,
    'date', date,
    'type', event_type,
    'attendees', attendees
)) from (
  select e.*, json_arrayagg(json_object(
      'id', a.id,
      'name', a.name,
      'email', a.email
  )) as attendees
  from events e
  left join (event_attendance ea, activists a)
  on (e.id = ea.event_id and ea.activist_id = a.id)
  where e.date >= '2019-12-01'
  group by e.id
) x;
```

I think we can also simplify the logic for handling where clauses by writing them like:

```
  where (:from_date = '' or e.date >= :from_date)
       and (:to_date = '' or e.date <= ':to_date')
```

I.e., rather than trying to build different SQL queries based on the search parameters, just write a query that works regardless and let the query optimizer handle it.

**Comment by mdempsky** [comment]
Handling event_type is a bit complex, because of "noConnections", "mpiDA", and "mpiCOM", but is still doable using CASE expressions, but not unreasonable. It could probably be simplified with help from a helper function. (Can't immediately make sense of how to use CREATE FUNCTION with MySQL though.)

Also, matching on events with a specified activist seemed like it would be complex, but is easily handled with a having clause:

```
  ...
  group by e.id
  having (:activist_name = '' or bit_or(a.name = :activist_name))
```

**Comment by jakehobbs** [comment]
Interesting. Didn't realize this was possible. Would updating or creating new events still be handled as it is now, or is there some other way to do that too?

**Comment by mdempsky** [comment]
Yeah, I didn't realize it was possible before seeing projects like PostGraphile and PostgREST.

As for mutation operations, I don't think there's any significantly new approach there, but I do think the API should be based on simpler atomic operations (eg, create event; create activist; mark activist as attending event; and so on) rather than large complex operations (eg, create this event and record these activists as attending and create any of them if they don't already exist too).
